diff --git a/src/nnn.c b/src/nnn.c
index 21a7370b..c0b4a238 100644
--- a/src/nnn.c
+++ b/src/nnn.c
@@ -3562,8 +3562,15 @@ static char *xreadline(const char *prefill, const char *prompt)
 	if (prefill) {
 		DPRINTF_S(prefill);
 		len = pos = mbstowcs(buf, prefill, READLINE_MAX);
-	} else
+
+		char *ext = strrchr(prefill, '.');
+		if (ext != NULL) {
+			size_t extlen = strlen(ext);
+			pos = len - extlen;
+		}
+	} else {
 		len = (size_t)-1;
+	}
 
 	if (len == (size_t)-1) {
 		buf[0] = '\0';
